adapters/sspbc/sspbc.go

```diff
@@ -0,0 +1,333 @@
+			var bidType openrtb_ext.BidType
+			var bidId = bid.ImpID
+
+			/*
+			  Determine bid type
+			  At this moment we only check if bid contains Adm property
+
+			  Later updates will check for video & native data
+			*/
+			if bid.AdM != "" {
+				bidType = openrtb_ext.BidTypeBanner
+			}
+
+			if BidExt, ok := a.adSlots[bidId]; ok {
+				var BidIdStored = BidExt.PbSlot
+				bid.ImpID = BidIdStored
+			}
+
+			// read additional data from proxy
+			var BidDataExt responseExt
+			if err := json.Unmarshal(bid.Ext, &BidDataExt); err != nil {
+				errors = append(errors, err)
+			} else {
+				var adCreationError error
+
+				/*
+					use correct ad creation method for a detected bid type
+					right now, we are only creating banner ads
+					if type is not detected / supported, throw error
+				*/
+				if bidType == openrtb_ext.BidTypeBanner {
+					bid.AdM, adCreationError = a.createBannerAd(bid, BidDataExt, internalRequest, seatBid.Seat)
+				} else {
+					adCreationError = fmt.Errorf("bid type is not supported: %s", bidType)
+				}
+
+				if adCreationError != nil {
+					errors = append(errors, err)
+				} else {
+					// append bid to responses
+					bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+						Bid:     &bid,
+						BidType: bidType,
+					})
+				}
+			}
+		}
+	}
+
+	return bidResponse, errors
+}
+
+func (a *adapter) createBannerAd(bid openrtb2.Bid, ext responseExt, request *openrtb2.BidRequest, seat string) (string, error) {
+	var mcad mcAd
+
+	if strings.Contains(bid.AdM, "<!--preformatted-->") {
+		// Banner ad is already formatted
+		return bid.AdM, nil
+	}
+
+	// create McAd payload
+	mcad.Id = request.ID
+	mcad.Seat = seat
+	mcad.SeatBid = make([]openrtb2.SeatBid, 1)
+	mcad.SeatBid[0].Bid = make([]openrtb2.Bid, 1)
+	mcad.SeatBid[0].Bid[0] = bid
+	mcMarshalled, err := json.Marshal(mcad)
+	if err != nil {
+		return bid.AdM, err
+	}
+
+	mcEncoded := base64.URLEncoding.EncodeToString(mcMarshalled)
+
+	bannerData := &templatePayload{
+		SiteId:  ext.SiteId,
+		SlotId:  ext.SlotId,
+		AdLabel: ext.AdLabel,
+		PubId:   ext.PublisherId,
+		Page:    request.Site.Page,
+		Referer: request.Site.Ref,
+		McAd:    mcEncoded,
+	}
+
+	// Prepare banner html, using template file
+	var filledTemplate bytes.Buffer
+	if err := a.bannerTemplate.Execute(&filledTemplate, bannerData); err != nil {
+		return bid.AdM, err
+	}
+
+	return filledTemplate.String(), nil
+}
+
+func getImpSize(Imp openrtb2.Imp) string {
+	impSize := "1x1"
+
+	if Imp.Banner != nil {
+		areaMax := int64(0)
+		for _, sizeI := range Imp.Banner.Format {
+			areaI := sizeI.W * sizeI.H
+			if areaI > areaMax {
+				areaMax = areaI
+				impSize = fmt.Sprintf("%dx%d", sizeI.W, sizeI.H)
+			}
+		}
+	}
+
+	// default fallback
+	return impSize
+}
+
+func (a *adapter) fillAdSlotData(impI openrtb2.Imp, impSize string) adSlotData {
+    var extData adSlotData
+    extData.PbSlot = impI.TagID
+    extData.PbSize = impSize
+    a.adSlots[impI.ID] = extData
```

Comment:
You are still reading and writing shared data in a thread unsafe way. Also you seem to be writing an entry for every impression, even if `a.adSlots[impI.ID]` already exists, which seems rather wasteful, especially if this interaction is made thread safe.

Reply:
OK, I've removed a.adSlots alltogether. What we used to do with it (request->response Imp matching) is done by looking at externalRequest